### PR TITLE
batchExecute functionality for SQL connector

### PR DESCRIPTION
**Pull request by niveathika**
## Purpose
Implement batchExecute functionality for SQL connector.
Partially fix https://github.com/ballerina-platform/ballerina-lang/issues/23565

## Approach
Used JDBC PreparedStatement batchExecute()

## Samples
```ballerina
import ballerina/io;
import ballerina/mysql;
import ballerina/sql;

public function main() {

    sql:Error? err = initializeClients();
    if (err is sql:Error) {
        io:println("Error occured", err);
    } else {
        io:println("Sample executed successfully!");
    }
}

function initializeClients(boolean test = false) returns sql:Error? {

    mysql:Client mysqlClient = check new ("localhost", "test", "", "test");
    io:println("MySQL client with user and password created.");

    _ = check mysqlClient->execute("DROP TABLE Persons");
    _ = check mysqlClient->execute("CREATE TABLE Persons (id int(11) NOT NULL AUTO_INCREMENT, name varchar(255) DEFAULT NULL, PRIMARY KEY (`id`))");

     sql:ParameterizedString sqlQuery1 = {
         parts : ["INSERT INTO Persons (name) values (",")"],
         insertions: ["Alice"]
     };

     sql:ParameterizedString sqlQuery2 = {
         parts : ["INSERT INTO Persons (name) values (",")"],
         insertions: ["Alice"]
     };

     sql:ParameterizedString sqlQuery3 = {
         parts : ["INSERT INTO Persons (name) values (",")"],
         insertions: ["Alice"]
     };

     sql:ParameterizedString[] sqlQueriues = [sqlQuery1, sqlQuery2, sqlQuery3];

    sql:ExecuteResult[]? result = check mysqlClient->batchExecute(sqlQueries);
    io:println(result);
    check mysqlClient.close();
}
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [x] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [x] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [x] Added necessary documentation  
  - [x] API documentation 
  - [ ] Module documentation in Module.md files
  - [x] Ballerina By Examples

### CODE REVIEW

File: stdlib/database/sql/src/main/java/org/ballerinalang/sql/utils/ExecuteUtils.java [hunk 1]
```diff
@@ -86,6 +93,118 @@ public static Object nativeExecute(ObjectValue client, MapValue<BString, Object>
         }
     }
 
+    public static Object nativeBatchExecute(ObjectValue client, ArrayValue paramSQLStrings,
+                                            boolean rollbackInFailure) {
+
```
Comment (sinthuja) [inline stdlib/database/sql/src/main/java/org/ballerinalang/sql/utils/ExecuteUtils.java:98]: Please remove unnecessary blank lines.

File: stdlib/database/sql/src/main/java/org/ballerinalang/sql/utils/ErrorGenerator.java [hunk 2]
```diff
@@ -54,6 +69,22 @@ public static ErrorValue getSQLApplicationError(String detailedErrorMessage) {
...
+        valueMap.put(Constants.ErrorRecordFields.EXECUTION_RESULTS,
+                BValueCreator.createArrayValue(executionResults.toArray(), new BArrayType(
+                  new BRecordType(Constants.EXECUTION_RESULT_RECORD, Constants.SQL_PACKAGE_ID, 0 , false, 0))));
+
+        MapValue<BString, Object> sqlClientErrorDetailRecord = BallerinaValues.
+                createRecordValue(Constants.SQL_PACKAGE_ID, Constants.BATCH_EXECUTE_ERROR_DATA, valueMap);
+        return BallerinaErrors.createError(Constants.BATCH_EXECUTE_ERROR_CODE, sqlClientErrorDetailRecord);
+    }
+
+
```
Comment (sinthuja) [inline stdlib/database/sql/src/main/java/org/ballerinalang/sql/utils/ErrorGenerator.java:87]: Please remove extra new line.

File: stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal [hunk 3]
```diff
@@ -94,6 +94,19 @@ public type Client client object {
         }
     }
 
+    public remote function batchExecute(sql:ParameterizedString[] sqlQueries, boolean rollbackInFailure = false)
```
Comment (sinthuja) [inline stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal:97]: Please add docs for the operation.

File: stdlib/database/sql/src/main/ballerina/src/sql/client.bal [hunk 4]
```diff
@@ -34,9 +34,12 @@ public type Client abstract client object {
     #
     # + sqlQuery - The DDL or DML query such as INSERT, DELETE, UPDATE, etc as `string` or `ParameterizedString`
     #              when the query has params to be passed in
-    # + return - Summary of the sql update query as `ExecuteResult` or returns `Error`
+    # + return - Summary of the sql update query as `ExecutionResult` or returns `Error`
     #           if any error occured when executing the query
-    public remote function execute(@untainted string|ParameterizedString sqlQuery) returns ExecuteResult|Error?;
+    public remote function execute(@untainted string|ParameterizedString sqlQuery) returns ExecutionResult|Error?;
+
+    public remote function batchExecute(ParameterizedString[] sqlQueries, boolean rollbackInFailure = false)
```
Comment (sinthuja) [inline stdlib/database/sql/src/main/ballerina/src/sql/client.bal:41]: Add method docs here as well.

File: stdlib/database/sql/src/main/ballerina/src/sql/error.bal [hunk 5]
```diff
@@ -14,6 +14,23 @@
 // specific language governing permissions and limitations
 // under the License.
 
+public const BATCH_EXECUTE_ERROR_REASON = "{ballerina/sql}BatchExecuteError";
```
Comment (sinthuja) [inline stdlib/database/sql/src/main/ballerina/src/sql/error.bal:17]: Missing docs.

File: stdlib/database/sql/src/main/ballerina/src/sql/error.bal [hunk 6]
```diff
@@ -14,6 +14,23 @@
...
+# + executionResults - Result of execution of commands.
+public type BatchExecuteErrorData record {|
+    string message?;
+    int errorCode;
+    string sqlState;
+    error cause?;
+    ExecutionResult[] executionResults;
+|};
+
 public const DATABASE_ERROR_REASON = "{ballerina/sql}DatabaseError";
```
Comment (sinthuja) [inline stdlib/database/sql/src/main/ballerina/src/sql/error.bal:34]: Missing docs.

File: stdlib/database/sql/src/main/ballerina/src/sql/error.bal [hunk 7]
```diff
@@ -44,6 +61,8 @@ public type ApplicationErrorData record {|
 # database resource clean-up, and other similar scenarios.
 public type DatabaseError error<DATABASE_ERROR_REASON, DatabaseErrorData>;
 
+public type BatchExecuteError error<BATCH_EXECUTE_ERROR_REASON, BatchExecuteErrorData>;
```
Comment (sinthuja) [inline stdlib/database/sql/src/main/ballerina/src/sql/error.bal:64]: missing docs

Comment (niveathika) [reply stdlib/database/sql/src/main/ballerina/src/sql/error.bal:65]: Fixed in c76f268

File: stdlib/database/sql/src/main/ballerina/src/sql/error.bal (hunk 6 above)
Comment (niveathika) [reply stdlib/database/sql/src/main/ballerina/src/sql/error.bal:34]: These const are used for internal error creation and will be removed after the error syntax migration is complete

File: stdlib/database/sql/src/main/ballerina/src/sql/client.bal (hunk 4 above)
Comment (niveathika) [reply stdlib/database/sql/src/main/ballerina/src/sql/client.bal:50]: Fixed in c76f268

File: stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal (hunk 3 above)
Comment (niveathika) [reply stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal:106]: Fixed in c76f268

File: stdlib/database/sql/src/main/ballerina/src/sql/error.bal (hunk 5 above)
Comment (niveathika) [reply stdlib/database/sql/src/main/ballerina/src/sql/error.bal:17]: These const are used for internal error creation and will be removed after the error syntax migration is complete

File: stdlib/database/sql/src/main/java/org/ballerinalang/sql/utils/ExecuteUtils.java (hunk 1 above)
Comment (niveathika) [reply stdlib/database/sql/src/main/java/org/ballerinalang/sql/utils/ExecuteUtils.java]: Fixed in 7bd6db9

File: stdlib/database/sql/src/main/java/org/ballerinalang/sql/utils/ErrorGenerator.java (hunk 2 above)
Comment (niveathika) [reply stdlib/database/sql/src/main/java/org/ballerinalang/sql/utils/ErrorGenerator.java]: Fixed in 7bd6db9

**Comment by sinthuja** [comment]
@niveathika , please send this PR to the master branch.

**Comment by niveathika** [comment]
@sinthuja sure. Moved to https://github.com/ballerina-platform/ballerina-lang/pull/23942 as there were some issues in checks

**Pull request closed without merging**